File: lib/alb_stack.py
"""
ALB Stack - Application Load Balancer with HTTPS Listener and S3 Logging
"""

import pulumi
import pulumi_aws as aws
from typing import Dict, List


class AlbStackArgs:
    """Arguments for AlbStack"""

    def __init__(
        self,
        environment_suffix: str,
        vpc_id: pulumi.Input[str],
        public_subnet_ids: List[pulumi.Input[str]],
        alb_sg_id: pulumi.Input[str],
        log_bucket_name: pulumi.Input[str],
        tags: Dict[str, str] = None
    ):
        self.environment_suffix = environment_suffix
        self.vpc_id = vpc_id
        self.public_subnet_ids = public_subnet_ids
        self.alb_sg_id = alb_sg_id
        self.log_bucket_name = log_bucket_name
        self.tags = tags or {}


class AlbStack(pulumi.ComponentResource):
    """
    Application Load Balancer for loan processing application.
    """

    def __init__(
        self,
        name: str,
        args: AlbStackArgs,
        opts: pulumi.ResourceOptions = None
    ):
        super().__init__("custom:alb:AlbStack", name, {}, opts)

        self.environment_suffix = args.environment_suffix
        self.tags = args.tags

        # Create Application Load Balancer
        self.alb = aws.lb.LoadBalancer(
            f"loan-alb-{self.environment_suffix}",
            name=f"loan-alb-{self.environment_suffix}",
            internal=False,
            load_balancer_type="application",
            security_groups=[args.alb_sg_id],
            subnets=args.public_subnet_ids,
            enable_deletion_protection=False,
            enable_http2=True,
            enable_cross_zone_load_balancing=True,
            access_logs=aws.lb.LoadBalancerAccessLogsArgs(
                bucket=args.log_bucket_name,
                enabled=True,
                prefix="alb-logs"
            ),
            tags={**self.tags, "Name": f"loan-alb-{self.environment_suffix}"},
            opts=pulumi.ResourceOptions(parent=self)
        )

        # Create Target Group
        self.target_group = aws.lb.TargetGroup(
            f"loan-tg-{self.environment_suffix}",
            name=f"loan-tg-{self.environment_suffix}",
            port=8080,
            protocol="HTTP",
            vpc_id=args.vpc_id,
            target_type="ip",
            deregistration_delay=30,
            health_check=aws.lb.TargetGroupHealthCheckArgs(
                enabled=True,
                healthy_threshold=2,
                interval=30,
                matcher="200",
                path="/",
                port="traffic-port",
                protocol="HTTP",
                timeout=5,
                unhealthy_threshold=3
            ),
            tags={**self.tags, "Name": f"loan-tg-{self.environment_suffix}"},
            opts=pulumi.ResourceOptions(parent=self)
        )

        # HTTP Listener (redirects to HTTPS)
        self.http_listener = aws.lb.Listener(
            f"loan-http-listener-{self.environment_suffix}",
            load_balancer_arn=self.alb.arn,
            port=80,
            protocol="HTTP",
            default_actions=[aws.lb.ListenerDefaultActionArgs(
                type="redirect",
                redirect=aws.lb.ListenerDefaultActionRedirectArgs(
                    port="443",
                    protocol="HTTPS",
                    status_code="HTTP_301"
                )
            )],
            tags={**self.tags, "Name": f"loan-http-listener-{self.environment_suffix}"},
            opts=pulumi.ResourceOptions(parent=self, depends_on=[self.alb])
        )

        # HTTPS Listener (using self-signed certificate for demo)
        # In production, use ACM certificate
        self.https_listener = aws.lb.Listener(
            f"loan-https-listener-{self.environment_suffix}",
            load_balancer_arn=self.alb.arn,
            port=443,
            protocol="HTTPS",
            ssl_policy="ELBSecurityPolicy-TLS13-1-2-2021-06",
            certificate_arn="arn:aws:acm:eu-west-2:123456789012:certificate/dummy",  # Replace with real ACM cert
            default_actions=[aws.lb.ListenerDefaultActionArgs(
                type="forward",
                target_group_arn=self.target_group.arn
            )],
            tags={**self.tags, "Name": f"loan-https-listener-{self.environment_suffix}"},
            opts=pulumi.ResourceOptions(
                parent=self,
                depends_on=[self.alb, self.target_group],
                ignore_changes=["certificate_arn"]
            )
        )

        self.register_outputs({
            "alb_arn": self.alb.arn,
            "alb_dns_name": self.alb.dns_name,
            "target_group_arn": self.target_group.arn
        })

File: lib/alb-stack.ts
/**
 * ALB Stack - Creates Application Load Balancer, target groups, and SSL certificates
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';
import { EnvironmentConfig, VpcOutputs, AlbOutputs } from './types';

export interface AlbStackArgs {
  config: EnvironmentConfig;
  vpcOutputs: VpcOutputs;
}

export class AlbStack extends pulumi.ComponentResource {
  public readonly outputs: AlbOutputs;

  constructor(
    name: string,
    args: AlbStackArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:alb:AlbStack', name, {}, opts);

    const { config, vpcOutputs } = args;

    // Create ALB security group
    const albSecurityGroup = new aws.ec2.SecurityGroup(
      `${config.environment}-alb-sg-${config.environmentSuffix}`,
      {
        vpcId: vpcOutputs.vpcId,
        description: `Security group for ${config.environment} ALB`,
        ingress: [
          {
            protocol: 'tcp',
            fromPort: 80,
            toPort: 80,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow HTTP traffic',
          },
          ...(config.enableSsl
            ? [
                {
                  protocol: 'tcp',
                  fromPort: 443,
                  toPort: 443,
                  cidrBlocks: ['0.0.0.0/0'],
                  description: 'Allow HTTPS traffic',
                },
              ]
            : []),
        ],
        egress: [
          {
            protocol: '-1',
            fromPort: 0,
            toPort: 0,
            cidrBlocks: ['0.0.0.0/0'],
            description: 'Allow all outbound traffic',
          },
        ],
        tags: {
          ...config.tags,
          Name: `${config.environment}-alb-sg-${config.environmentSuffix}`,
        },
      },
      { parent: this }
    );

    // Create Application Load Balancer
    const alb = new aws.lb.LoadBalancer(
      `${config.environment}-alb-${config.environmentSuffix}`,
      {
        internal: false,
        loadBalancerType: 'application',
        securityGroups: [albSecurityGroup.id],
        subnets: vpcOutputs.publicSubnetIds,
        enableDeletionProtection: false,
        tags: {
          ...config.tags,
          Name: `${config.environment}-alb-${config.environmentSuffix}`,
        },
      },
      { parent: this }
    );

    // Create target group
    const targetGroup = new aws.lb.TargetGroup(
      `${config.environment}-tg-${config.environmentSuffix}`,
      {
        port: 3000,
        protocol: 'HTTP',
        vpcId: vpcOutputs.vpcId,
        targetType: 'ip',
        healthCheck: {
          enabled: true,
          path: '/health',
          protocol: 'HTTP',
          matcher: '200',
          interval: 30,
          timeout: 5,
          healthyThreshold: 2,
          unhealthyThreshold: 3,
        },
        deregistrationDelay: 30,
        tags: {
          ...config.tags,
          Name: `${config.environment}-tg-${config.environmentSuffix}`,
        },
      },
      { parent: this }
    );

    let albUrl: pulumi.Output<string>;

    if (config.enableSsl) {
      // Create ACM certificate for HTTPS
      const certificate = new aws.acm.Certificate(
        `${config.environment}-cert-${config.environmentSuffix}`,
        {
          domainName: `${config.environment}.example.com`,
          validationMethod: 'DNS',
          tags: {
            ...config.tags,
            Name: `${config.environment}-cert-${config.environmentSuffix}`,
          },
        },
        { parent: this }
      );

      // Create HTTPS listener
      new aws.lb.Listener(
        `${config.environment}-https-listener-${config.environmentSuffix}`,
        {
          loadBalancerArn: alb.arn,
          port: 443,
          protocol: 'HTTPS',
          sslPolicy: 'ELBSecurityPolicy-TLS-1-2-2017-01',
          certificateArn: certificate.arn,
          defaultActions: [
            {
              type: 'forward',
              targetGroupArn: targetGroup.arn,
            },
          ],
          tags: {
            ...config.tags,
            Name: `${config.environment}-https-listener-${config.environmentSuffix}`,
          },
        },
        { parent: this }
      );

      // HTTP listener redirects to HTTPS
      new aws.lb.Listener(
        `${config.environment}-http-listener-${config.environmentSuffix}`,
        {
          loadBalancerArn: alb.arn,
          port: 80,
          protocol: 'HTTP',
          defaultActions: [
            {
              type: 'redirect',
              redirect: {
                port: '443',
                protocol: 'HTTPS',
                statusCode: 'HTTP_301',
              },
            },
          ],
          tags: {
            ...config.tags,
            Name: `${config.environment}-http-listener-${config.environmentSuffix}`,
          },
        },
        { parent: this }
      );

      albUrl = pulumi.interpolate`https://${alb.dnsName}`;
    } else {
      // Create HTTP listener only
      new aws.lb.Listener(
        `${config.environment}-http-listener-${config.environmentSuffix}`,
        {
          loadBalancerArn: alb.arn,
          port: 80,
          protocol: 'HTTP',
          defaultActions: [
            {
              type: 'forward',
              targetGroupArn: targetGroup.arn,
            },
          ],
          tags: {
            ...config.tags,
            Name: `${config.environment}-http-listener-${config.environmentSuffix}`,
          },
        },
        { parent: this }
      );

      albUrl = pulumi.interpolate`http://${alb.dnsName}`;
    }

    this.outputs = {
      albArn: alb.arn,
      albDnsName: alb.dnsName,
      albUrl: albUrl,
      targetGroupArn: targetGroup.arn,
      securityGroupId: albSecurityGroup.id,
    };

    this.registerOutputs({
      albArn: this.outputs.albArn,
      albDnsName: this.outputs.albDnsName,
      albUrl: this.outputs.albUrl,
      targetGroupArn: this.outputs.targetGroupArn,
      securityGroupId: this.outputs.securityGroupId,
    });
  }
}
